# Data Literacy, Pandas Library & Exploration Data Analysis(EDA)

In this section we will cover the topics such as; how to handle and process the data, how to read it, how to visualize the data, what can we deduct from a dataset after the processes that we made and at the end we apply EDA(Exploration Data Analysis) on a valid dataset.

* [Data Types](#datatypes)
* [Measures of Central Tendency](#measures_of_central_tendency)
    * [Measures of Central Tendency In Practice](#measures_of_central_tendency_in_practice)
* [Measures of Dispersion](#measures_of_dispersion)
    * [Measures of Dispersion In Practice](#measures_of_dispersion_in_practice)
* [Essentials of Pandas](#pandas)
    * [Data Frames and Series](#dfs)
    * [Indexing a Data Frame](#index)
    * [Selecting a Specific Part of a Data Frame (iloc and loc)](#iloc_loc)
    * [Modfying Columns and Adding New Columns](#cols)
    * [Resetting the Data Frame](#resetting)
    * [cut and qcut Methods of Pandas](#cutting)
    * [Deleting Rows and Columns](#del_col)
    * [Missing Values](#missing)

<a id="datatypes"></a>
# Data Types

* **Qualitative Variable (Categorical Variables)**
  <br>Qualitative data represents the data type that can be categorized:
    * Gender
    * Color
    * Breed
    * etc...
* **Quantitative Variables (Numerical Variables)**
  <br>Quantitative data represents the data type that arithmetic operators can be applied on
    * Age
    * Income
    * Temperature
    * etc...

_____

These **categorical variables** and **numerical variables** branch into two different categories:

* **Categorical Variables**
    * Nominal
    <br>Nominal variables categorizes itself without any order, such as:
        * Gender
        * Color
        * etc...
    * Ordinal
    <br>Ordinal variables categorizes itself with a specific order, such as:
        * Grade
        * Educational Status
        * Military Rank
        * etc...
* **Numerical Variables**
    * Interval
    <br>In interval scale; numerical differences have a meaning, arithmetic operators can be applied on and there is no absolute zero. Such as:
        * Celcius (C)
        * Fahrenheit (F)
        * etc...
    * Ratio
    <br>In interval scale; numerical differences have a meaning, arithmetic operators can be applied on but there is an absolute zero. Such as:
        * Age
        * Length
        * etc...
_____

In summary;

**Data** branches into two:

* **Qualitative Variable (Categorical Variables)**
    * Nominal
    * Ordinal
* **Quantitative Variables (Numerical Variables)**
    * Interval
    * Ratio

<a id="measures_of_central_tendency"></a>
# Measures of Central Tendency

Measures of central tendency is central statistical values of a data distribution. The purpose of it to comprehend and summarize a dataset.

To calculate such measurements in python, we benefit from some libraries like: `numpy` ,`pandas`, `scipy`...

For now, `numpy` and `pandas` are enough.

* **Mean**<br>
The sum of all measurements divided by the number of observations in the data set.
* **Median**<br>
The middle value that separates the higher half from the lower half of the data set.
* **Mode**<br>
The most frequent value in the data set.
* **Quartiles**<br>
Type of quantiles which divide the number of data points into four parts, or quarters.




<a id="measures_of_central_tendency_in_practice"></a>
# Measures of Central Tendency In Practice

In [2]:
# Required Libraries
import numpy as np
import pandas as pd

In [3]:
# Creating an array contains random integers
random_array = np.random.randint(10, size=(10))
random_array

array([6, 1, 9, 4, 7, 3, 9, 2, 9, 9])

In [4]:
# Creating a dataframe to apply measurements
s = pd.DataFrame(random_array)

In [5]:
# Mean Value
s.mean()

0    5.9
dtype: float64

In [6]:
# Median Value
s.median()

0    6.5
dtype: float64

In [7]:
# Mode Value
s.mode()

0
0  9

In [8]:
# Quartiles
s.quantile(q=[0.25, 0.50, 0.75])

0
0.25  3.25
0.50  6.50
0.75  9.00

<a id="measures_of_dispersion"></a>
# Measures of Dispersion

Measures of dispersion is the extent to which a distribution is stretched or squeezed. Common examples of measures of statistical dispersion are the variance, standard deviation, and interquartile range.

* **Range**<br>
The range of a set of data is size of the narrowest interval which contains all the data.
* **Standard Deviation**<br>
The standard deviation is a measure of the amount of variation of the values of a variable about its mean.
* **Variance**<br>
The expected value of the squared deviation from the mean of a random variable.
* **Interquartile Range(IQR)**<br>
The interquartile range (IQR) is a measure of statistical dispersion, which is the spread of the data.

<a id="measures_of_dispersion_in_practice"></a>
# Measures of Dispersion In Practice

In [9]:
# Range
range = s.max() - s.min()
range

0    8
dtype: int64

In [10]:
# Standard Deviation
s.std()

0    3.17805
dtype: float64

In [11]:
# Variance
s.var()

0    10.1
dtype: float64

In [12]:
# Quartiles
quartiles = s.quantile(q=[0.25, 0.50, 0.75])
# IQR
IQR = quartiles.iloc[2] - quartiles.iloc[0]
IQR

0    5.75
dtype: float64

<a id="pandas"></a>
# Essentials of Pandas Library

In [13]:
# Importing required libraries
import numpy as np
import pandas as pd
import seaborn as sns # to import a dataset and visualize the dataset
import matplotlib.pyplot as plt # to visualize the dataset

<a id="dfs"></a>
## Data Frames and Series

First, we should create a dataframe or series to `use pandas essentially`. Dataframes and series can be created by the user or from an available dataset.

Main `difference between dataframes and series` is dimensions of them. Series has 1 dimension yet dataframe has 2. This explanation makes more sense as the section progress.

### Creating Series

In [14]:
# Creating a Series with a List object
s = pd.Series(data=[1, 2, 3, 4, 5, 6, 7, 8, 9])
s

0    1
1    2
2    3
3    4
4    5
5    6
6    7
7    8
8    9
dtype: int64

In [15]:
array = np.random.randint(100, size=(10))

# Creating a Series from a numpy array
s = pd.Series(array)
s

0    66
1    17
2    45
3    45
4    65
5    76
6    87
7    29
8    62
9    29
dtype: int64

In [16]:
dict_a = {"A": 1, "B": 2, "C": 3, "D": 4}

# Creating a series with a dictionary
s = pd.Series(dict_a)
s

A    1
B    2
C    3
D    4
dtype: int64

In [17]:
list_index = ["A", "B", "C", "D"]
list_data = [1, 2, 3, 4]

# Creating a series with two separate lists
s = pd.Series(data=list_data, index=list_index)
s

A    1
B    2
C    3
D    4
dtype: int64

### Creating Dataframes

A dataframe can easily be created by reading a file(.xlsx, .csv, .json, .sql, .pickle, .txt) as long as the **correct path** is given:

    df = pd.read_excel("data.xlsx")
    df = pd.read_csv("data.csv")
    df = pd.read_json("data.json")
    df = pd.read_sql("data.sql")
    df = pd.read_pickle("data.pickle")
    df = pd.read_csv("data.txt")

#### Creating Dataframe with Seaborn

In [18]:
# Importing a dataset from seaborn library
titanic_dataset = sns.load_dataset("titanic")

# Creating a DataFrame from the copy of the dataset
df = titanic_dataset.copy()

# We call `head()` function to check the *df* dataframe
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

|  *Features*=>  |   survived |   pclass | sex    |   age |   sibsp |   parch |    fare | embarked   | class   | who   | adult_male   | deck   | embark_town   | alive   | alone   |
|---:|-----------:|---------:|:-------|------:|--------:|--------:|--------:|:-----------|:--------|:------|:-------------|:-------|:--------------|:--------|:--------|
|  Results of one sample=> |          0 |        3 | male   |    22 |       1 |       0 |  7.25   | S          | Third   | man   | True         | nan    | Southampton   | no      | False   |
|  1 |          1 |        1 | female |    38 |       1 |       0 | 71.2833 | C          | First   | woman | False        | C      | Cherbourg     | yes     | False   |
|  2 |          1 |        3 | female |    26 |       0 |       0 |  7.925  | S          | Third   | woman | False        | nan    | Southampton   | yes     | True    |
|  3 |          1 |        1 | female |    35 |       1 |       0 | 53.1    | S          | First   | woman | False        | C      | Southampton   | yes     | False   |
|  4 |          0 |        3 | male   |    35 |       0 |       0 |  8.05   | S          | Third   | man   | True         | nan    | Southampton   | no      | True    |

<a id="index"></a>
## Indexing a Data Frame

### Choosing Single Feature

In [19]:
# choosing 'age' as series from df
df["age"]

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: age, Length: 891, dtype: float64

In [20]:
# another method to choose 'age' as series from df
df.age

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: age, Length: 891, dtype: float64

In [21]:
# choosing 'age' as a dataframe from df
df[["age"]]

age
0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
..    ...
886  27.0
887  19.0
888   NaN
889  26.0
890  32.0

[891 rows x 1 columns]

**Important Note:**

As it can be easily noticed; a feature can be called with many different methods. Main difference of using double brackets: `[["age"]]` is to define the called object as a dataframe. Also if we want to call multiple features, we have to call it as a dataframe: `[[]]` not as a series `[]`

In [22]:
# Choosing as series
print(type(df["age"]))

# Choosing as a data frame
print(type(df[["age"]]))

<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>


### Choosing Multiple Features

In [23]:
df[["age", "sex"]]

age     sex
0    22.0    male
1    38.0  female
2    26.0  female
3    35.0  female
4    35.0    male
..    ...     ...
886  27.0    male
887  19.0  female
888   NaN  female
889  26.0    male
890  32.0    male

[891 rows x 2 columns]

### Fancy Indexing

We write an array inside of our `df` to call mutliple features, we can do the same thing by giving an array that is created outside of it.

In [24]:
# a list that contains our columns that we want to observe
columns = ["survived", "sex", "age"]

# fancy indexing
df[columns]

survived     sex   age
0           0    male  22.0
1           1  female  38.0
2           1  female  26.0
3           1  female  35.0
4           0    male  35.0
..        ...     ...   ...
886         0    male  27.0
887         1  female  19.0
888         0  female   NaN
889         1    male  26.0
890         0    male  32.0

[891 rows x 3 columns]

<a id="iloc_loc"></a>
## Selecting Specific Parts of a Data Frame via **.iloc** and **.loc**

Specific parts of a data frame can be selected with the methods of pandas called **iloc** and **loc**. It is similar to the base slicing of Python

**iloc:**

Specific parts of a data frame can be selected with **iloc** via slicing the data frame with integers.

**loc:**

With loc on the other hand; specific parts can be selected with the conditions defined in `loc` method. (The data will be returned as the condition is met.)

### Index Based Selection (iloc Method)

It uses integers to slice the data frame

In [25]:
# Let's check our data frame before using iloc method
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [26]:
df.iloc[0, 2] # this expression selects the 0 => first row, 2 => third column

'male'

In [27]:
df.iloc[:, 0] # : => all rows, 0 => first column

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: survived, Length: 891, dtype: int64

In [28]:
df.iloc[0, :] # 0 => first row, : => all columns

survived                 0
pclass                   3
sex                   male
age                   22.0
sibsp                    1
parch                    0
fare                  7.25
embarked                 S
class                Third
who                    man
adult_male            True
deck                   NaN
embark_town    Southampton
alive                   no
alone                False
Name: 0, dtype: object

In [29]:
df.iloc[:, :] # all rows, all columns

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [30]:
df.iloc[:: -1] # all rows and all columns but reversed

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
890         0       3    male  32.0      0      0   7.7500        Q   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
887         1       1  female  19.0      0      0  30.0000        S   First   
886         0       2    male  27.0      0      0  13.0000        S  Second   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
4           0       3    male  35.0      0      0   8.0500        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
0           0       3    male  22.0      1      0   7.2500        S   Third   

       who  adult_male deck  embark_town alive  alone  
890    man        True  NaN   Queenstown    no   True  
889    man        True    C    Cherbourg   yes   True  
888  woman       False  NaN  Southampton    no  False  
887  woman       False    B  Southampton   yes   True  
886    man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
4      man        True  NaN  Southampton    no   True  
3    woman       False    C  Southampton   yes  False  
2    woman       False  NaN  Southampton   yes   True  
1    woman       False    C    Cherbourg   yes  False  
0      man        True  NaN  Southampton    no  False  

[891 rows x 15 columns]

In [31]:
df.iloc[::2] # all rows that has even index numbers and all columns

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
2           1       3  female  26.0      0      0   7.9250        S   Third   
4           0       3    male  35.0      0      0   8.0500        S   Third   
6           0       1    male  54.0      0      0  51.8625        S   First   
8           1       3  female  27.0      0      2  11.1333        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
882         0       3  female  22.0      0      0  10.5167        S   Third   
884         0       3    male  25.0      0      0   7.0500        S   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
888         0       3  female   NaN      1      2  23.4500        S   Third   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
2    woman       False  NaN  Southampton   yes   True  
4      man        True  NaN  Southampton    no   True  
6      man        True    E  Southampton    no   True  
8    woman       False  NaN  Southampton   yes  False  
..     ...         ...  ...          ...   ...    ...  
882  woman       False  NaN  Southampton    no   True  
884    man        True  NaN  Southampton    no   True  
886    man        True  NaN  Southampton    no   True  
888  woman       False  NaN  Southampton    no  False  
890    man        True  NaN   Queenstown    no   True  

[446 rows x 15 columns]

In [32]:
df.iloc[0:3, 0:3] # first three rows and first three columns

survived  pclass     sex
0         0       3    male
1         1       1  female
2         1       3  female

### Label Based Selection (loc method)

It uses feature names and conditions to slice a data frame

In [33]:
df.loc[:, "age"] # all rows of `age` column

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: age, Length: 891, dtype: float64

In [34]:
df.loc[df["age"] > 40] # all columns where `age` is greater than 40

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
6           0       1    male  54.0      0      0  51.8625        S   First   
11          1       1  female  58.0      0      0  26.5500        S   First   
15          1       2  female  55.0      0      0  16.0000        S  Second   
33          0       2    male  66.0      0      0  10.5000        S  Second   
35          0       1    male  42.0      1      0  52.0000        S   First   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
862         1       1  female  48.0      0      0  25.9292        S   First   
865         1       2  female  42.0      0      0  13.0000        S  Second   
871         1       1  female  47.0      1      1  52.5542        S   First   
873         0       3    male  47.0      0      0   9.0000        S   Third   
879         1       1  female  56.0      0      1  83.1583        C   First   

       who  adult_male deck  embark_town alive  alone  
6      man        True    E  Southampton    no   True  
11   woman       False    C  Southampton   yes   True  
15   woman       False  NaN  Southampton   yes   True  
33     man        True  NaN  Southampton    no   True  
35     man        True  NaN  Southampton    no  False  
..     ...         ...  ...          ...   ...    ...  
862  woman       False    D  Southampton   yes   True  
865  woman       False  NaN  Southampton   yes   True  
871  woman       False    D  Southampton   yes  False  
873    man        True  NaN  Southampton    no   True  
879  woman       False    C    Cherbourg   yes  False  

[150 rows x 15 columns]

In [35]:
df.loc[df["age"] > 40, ["survived", "sex", "fare"]] # survived, sex, and fare columns where the `age > 40` condition is met

survived     sex     fare
6           0    male  51.8625
11          1  female  26.5500
15          1  female  16.0000
33          0    male  10.5000
35          0    male  52.0000
..        ...     ...      ...
862         1  female  25.9292
865         1  female  13.0000
871         1  female  52.5542
873         0    male   9.0000
879         1  female  83.1583

[150 rows x 3 columns]

In [36]:
# `&`, `|` operators can be used to apply multiple conditions in selection

df.loc[(df["age"] > 40)
        & (df["sex"]=="male")
        & ((df["embark_town"]=="Southampton") | (df["embark_town"]=="Cherbourg")),
        ["survived", "pclass", "age", "fare"]]
# we want to observe `survived, pclass, age, fare` columns where age is greater than 40 `AND` sex is male `AND` embark town is Southampton `OR` Cherbourg

survived  pclass   age     fare
6           0       1  54.0  51.8625
33          0       2  66.0  10.5000
35          0       1  42.0  52.0000
54          0       1  65.0  61.9792
62          0       1  45.0  83.4750
..        ...     ...   ...      ...
845         0       3  42.0   7.5500
851         0       3  74.0   7.7750
857         1       1  51.0  26.5500
860         0       3  41.0  14.1083
873         0       3  47.0   9.0000

[97 rows x 4 columns]

In [37]:
df.loc[(df["age"] > 40)
       & (df["sex"] == "male")
       & ((df["embark_town"] == "Cherbourg") | (df["embark_town"] == "Southampton")),
       ["age", "class", "embark_town"]]

age   class  embark_town
6    54.0   First  Southampton
33   66.0  Second  Southampton
35   42.0   First  Southampton
54   65.0   First    Cherbourg
62   45.0   First  Southampton
..    ...     ...          ...
845  42.0   Third  Southampton
851  74.0   Third  Southampton
857  51.0   First  Southampton
860  41.0   Third  Southampton
873  47.0   Third  Southampton

[97 rows x 3 columns]

In [57]:
df.loc[:, df.columns.str.contains("age")].head()  # selecting the values that has `age` in it's name

age
0  22.0
1  38.0
2  26.0
3  35.0
4  35.0

In [58]:
df.loc[:, ~df.columns.str.contains("age")].head()  # selecting the values that DOES NOT have `age` in it's name

survived  pclass     sex  sibsp  parch     fare embarked  class    who  \
0         0       3    male      1      0   7.2500        S  Third    man   
1         1       1  female      1      0  71.2833        C  First  woman   
2         1       3  female      0      0   7.9250        S  Third  woman   
3         1       1  female      1      0  53.1000        S  First  woman   
4         0       3    male      0      0   8.0500        S  Third    man   

   adult_male deck  embark_town alive  alone  
0        True  NaN  Southampton    no  False  
1       False    C    Cherbourg   yes  False  
2       False  NaN  Southampton   yes   True  
3       False    C  Southampton   yes  False  
4        True  NaN  Southampton    no   True

<a id="cols"></a>
## Modifying Columns and Adding New Columns

Current columns can be modified or even new ones can be added.

    df["new_column_name"] = new_condition
    df["currrent_column] = new_condition

Let's create a `new age` column that is substracted from the current year:

In [38]:
import datetime

current_date = datetime.date.today()
year = current_date.strftime("%Y")
year = int(year)
year

2024

In [39]:
df["new_age"] = year - df["age"]
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  new_age  
0    man        True  NaN  Southampton    no  False   2002.0  
1  woman       False    C    Cherbourg   yes  False   1986.0  
2  woman       False  NaN  Southampton   yes   True   1998.0  
3  woman       False    C  Southampton   yes  False   1989.0  
4    man        True  NaN  Southampton    no   True   1989.0

### Setting a Column as Index

In [40]:
# to set it as an index in our dataframe we need `set_index` method and set `inplace` parameter as True to apply it on df
df.set_index("new_age", inplace=True)
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  \
new_age                                                                   
2002.0          0       3    male  22.0      1      0   7.2500        S   
1986.0          1       1  female  38.0      1      0  71.2833        C   
1998.0          1       3  female  26.0      0      0   7.9250        S   
1989.0          1       1  female  35.0      1      0  53.1000        S   
1989.0          0       3    male  35.0      0      0   8.0500        S   

         class    who  adult_male deck  embark_town alive  alone  
new_age                                                           
2002.0   Third    man        True  NaN  Southampton    no  False  
1986.0   First  woman       False    C    Cherbourg   yes  False  
1998.0   Third  woman       False  NaN  Southampton   yes   True  
1989.0   First  woman       False    C  Southampton   yes  False  
1989.0   Third    man        True  NaN  Southampton    no   True

In [41]:
# we can also change the name of the index
df.index.names = ["new index"]
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  \
new index                                                                   
2002.0            0       3    male  22.0      1      0   7.2500        S   
1986.0            1       1  female  38.0      1      0  71.2833        C   
1998.0            1       3  female  26.0      0      0   7.9250        S   
1989.0            1       1  female  35.0      1      0  53.1000        S   
1989.0            0       3    male  35.0      0      0   8.0500        S   

           class    who  adult_male deck  embark_town alive  alone  
new index                                                           
2002.0     Third    man        True  NaN  Southampton    no  False  
1986.0     First  woman       False    C    Cherbourg   yes  False  
1998.0     Third  woman       False  NaN  Southampton   yes   True  
1989.0     First  woman       False    C  Southampton   yes  False  
1989.0     Third    man        True  NaN  Southampton    no   True

<a id="resetting"></a>
## Resetting the Data Frame

In [42]:
# Sometimes we have to reset the data frame because it becomes unusable as we mess around with it ,but; we can reset the df as we have a copy of the main data frame as `titanic_dataset` variable
df = titanic_dataset.copy()
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

<a id="cutting"></a>
## cut and qcut Methods of Pandas

Pandas have `cut` and `qcut` methods when we need to cut a column up in parts to categorize it.

### **cut Method**
**Parameters:**
* **x:** the data to be cut
* **bins:** main condition to cut
* **labels:** assigns labels for the cut (uses conditions if not defined)
* **right:** decides whether highest value is included or not
* **include_lowest:** self explanatory

____

### **qcut Method**

The main difference from `cut` is that `qcut` divides the selected column as given quantiles.

**Parameters:**
* **x:** the data to be cut
* **q:** the quantiles as a list
* **labels:** assigns labels for the cut (uses conditions if not defined)


In [43]:
# cut method on practice
pd.cut(df["age"], [0, 10, 18, 25, 40, 90], labels = ["child", "adolescence", "young-adult", "adult", "senior"])

0      young-adult
1            adult
2            adult
3            adult
4            adult
          ...     
886          adult
887    young-adult
888            NaN
889          adult
890          adult
Name: age, Length: 891, dtype: category
Categories (5, object): ['child' < 'adolescence' < 'young-adult' < 'adult' < 'senior']

In [44]:
# qcut method on practice
pd.qcut(df["age"], q=[0, 0.20, 0.40, 0.60, 0.80, 1], labels = ["child", "adolescence", "young-adult", "adult", "senior"])

0      adolescence
1            adult
2      young-adult
3            adult
4            adult
          ...     
886    young-adult
887          child
888            NaN
889    young-adult
890          adult
Name: age, Length: 891, dtype: category
Categories (5, object): ['child' < 'adolescence' < 'young-adult' < 'adult' < 'senior']

<a id="del_col"></a>
## Deleting Rows and Columns
Essential part of this process is `"axis"`:
* `"axis"=0` means rows
* `"axis"=1` means columns

### Deleting Columns

In [45]:
df.drop("fare", axis=1) # deletes `fare` column temporarily

survived  pclass     sex   age  sibsp  parch embarked   class    who  \
0           0       3    male  22.0      1      0        S   Third    man   
1           1       1  female  38.0      1      0        C   First  woman   
2           1       3  female  26.0      0      0        S   Third  woman   
3           1       1  female  35.0      1      0        S   First  woman   
4           0       3    male  35.0      0      0        S   Third    man   
..        ...     ...     ...   ...    ...    ...      ...     ...    ...   
886         0       2    male  27.0      0      0        S  Second    man   
887         1       1  female  19.0      0      0        S   First  woman   
888         0       3  female   NaN      1      2        S   Third  woman   
889         1       1    male  26.0      0      0        C   First    man   
890         0       3    male  32.0      0      0        Q   Third    man   

     adult_male deck  embark_town alive  alone  
0          True  NaN  Southampton    no  False  
1         False    C    Cherbourg   yes  False  
2         False  NaN  Southampton   yes   True  
3         False    C  Southampton   yes  False  
4          True  NaN  Southampton    no   True  
..          ...  ...          ...   ...    ...  
886        True  NaN  Southampton    no   True  
887       False    B  Southampton   yes   True  
888       False  NaN  Southampton    no  False  
889        True    C    Cherbourg   yes   True  
890        True  NaN   Queenstown    no   True  

[891 rows x 14 columns]

In [46]:
df.drop("fare", axis=1, inplace=True) # deletes `fare` column permanently
df

survived  pclass     sex   age  sibsp  parch embarked   class    who  \
0           0       3    male  22.0      1      0        S   Third    man   
1           1       1  female  38.0      1      0        C   First  woman   
2           1       3  female  26.0      0      0        S   Third  woman   
3           1       1  female  35.0      1      0        S   First  woman   
4           0       3    male  35.0      0      0        S   Third    man   
..        ...     ...     ...   ...    ...    ...      ...     ...    ...   
886         0       2    male  27.0      0      0        S  Second    man   
887         1       1  female  19.0      0      0        S   First  woman   
888         0       3  female   NaN      1      2        S   Third  woman   
889         1       1    male  26.0      0      0        C   First    man   
890         0       3    male  32.0      0      0        Q   Third    man   

     adult_male deck  embark_town alive  alone  
0          True  NaN  Southampton    no  False  
1         False    C    Cherbourg   yes  False  
2         False  NaN  Southampton   yes   True  
3         False    C  Southampton   yes  False  
4          True  NaN  Southampton    no   True  
..          ...  ...          ...   ...    ...  
886        True  NaN  Southampton    no   True  
887       False    B  Southampton   yes   True  
888       False  NaN  Southampton    no  False  
889        True    C    Cherbourg   yes   True  
890        True  NaN   Queenstown    no   True  

[891 rows x 14 columns]

### Deleting Rows

In [47]:
index_list = [0, 2, 4, 6]
df.drop(index_list, axis=0)  # deletes the defined index list

survived  pclass     sex   age  sibsp  parch embarked   class    who  \
1           1       1  female  38.0      1      0        C   First  woman   
3           1       1  female  35.0      1      0        S   First  woman   
5           0       3    male   NaN      0      0        Q   Third    man   
7           0       3    male   2.0      3      1        S   Third  child   
8           1       3  female  27.0      0      2        S   Third  woman   
..        ...     ...     ...   ...    ...    ...      ...     ...    ...   
886         0       2    male  27.0      0      0        S  Second    man   
887         1       1  female  19.0      0      0        S   First  woman   
888         0       3  female   NaN      1      2        S   Third  woman   
889         1       1    male  26.0      0      0        C   First    man   
890         0       3    male  32.0      0      0        Q   Third    man   

     adult_male deck  embark_town alive  alone  
1         False    C    Cherbourg   yes  False  
3         False    C  Southampton   yes  False  
5          True  NaN   Queenstown    no   True  
7         False  NaN  Southampton    no  False  
8         False  NaN  Southampton   yes  False  
..          ...  ...          ...   ...    ...  
886        True  NaN  Southampton    no   True  
887       False    B  Southampton   yes   True  
888       False  NaN  Southampton    no  False  
889        True    C    Cherbourg   yes   True  
890        True  NaN   Queenstown    no   True  

[887 rows x 14 columns]

In [51]:
# resetting the df
df = titanic_dataset.copy()
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

<a id="missing"></a>
## Missing Values

### Calculating the Missing Values

    df.isnull() # returns boolean values; `False` for correct data, `True` for missing data
    df.isnull().sum() # returns total count of missing values for each column
    df.isnull().sum().sum() # returns total count of missing values for whole data frame

In [52]:
df.isnull()

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0       False   False  False  False  False  False  False     False  False   
1       False   False  False  False  False  False  False     False  False   
2       False   False  False  False  False  False  False     False  False   
3       False   False  False  False  False  False  False     False  False   
4       False   False  False  False  False  False  False     False  False   
..        ...     ...    ...    ...    ...    ...    ...       ...    ...   
886     False   False  False  False  False  False  False     False  False   
887     False   False  False  False  False  False  False     False  False   
888     False   False  False   True  False  False  False     False  False   
889     False   False  False  False  False  False  False     False  False   
890     False   False  False  False  False  False  False     False  False   

       who  adult_male   deck  embark_town  alive  alone  
0    False       False   True        False  False  False  
1    False       False  False        False  False  False  
2    False       False   True        False  False  False  
3    False       False  False        False  False  False  
4    False       False   True        False  False  False  
..     ...         ...    ...          ...    ...    ...  
886  False       False   True        False  False  False  
887  False       False  False        False  False  False  
888  False       False   True        False  False  False  
889  False       False  False        False  False  False  
890  False       False   True        False  False  False  

[891 rows x 15 columns]

In [55]:
df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [56]:
df.isnull().sum().sum()

869

### Handling Missing Values

#### Deleting Missing Values

    df.dropna()  # Deletes every row that has NaN value, defult value for axis is 0
    df.dropna(axis=1)  # Deletes every column that has NaN value if axis is entered as 1
    df.dropna(thresh=2)  # Deletes every row that has at least 2 NaN values in it

#### Filling Missing Values

    df.fillna(value=1)  # Fills missing values with 1

    def mean_val(df, column):
      return int(df[column].mean())

    df["age"].fillna(value=mean_val(df, "age")) #fills missing values with mean value

# Exploratory Data Analysis(EDA)

## General Information About a Dataset

* Head and tail of the dataset
* Shape and size of the dataset
* Data types of the dataset
* Index pattern of the dataset
* Number of dimensions that the dataset has
* Statistical description of the dataset